README.rst
=======================
neutron-lbaas-dashboard
=======================

Horizon panels for Neutron LBaaS v2

* Free software: Apache license
* Documentation: http://docs.openstack.org/developer/neutron-lbaas-dashboard
* Source: http://git.openstack.org/cgit/openstack/neutron-lbaas-dashboard
* Bugs: http://bugs.launchpad.net/neutron

Features
--------

* Please see neutron-lbaas repository


Howto
-----

1. Package the f5_lbaas_dashboard by running::

    python setup.py sdist

   This will create a python egg in the dist folder, which can be used to
   install on the horizon machine or within horizon's python virtual
   environment.

2. Copy ``_1481_project_ng_loadbalancersv2_panel.py`` in
   ``f5_lbaas_dashboard/enabled`` directory
   to ``openstack_dashboard/local/enabled``.

3. (Optional) Copy the policy file into horizon's policy files folder, and
   add this config ``POLICY_FILES``::

    'neutron_lbaas': 'neutron_lbaas_policy.json',

4. Django has a compressor feature that performs many enhancements for the
   delivery of static files. If the compressor feature is enabled in your
   environment (``COMPRESS_OFFLINE = True``), run the following commands::

    $ ./manage.py collectstatic
    $ ./manage.py compress

5. Finally restart your web server to enable neutron-lbaas-dashboard
   in your Horizon::

    $ sudo service apache2 restart


ChangeLog
CHANGES
=======

* change all neutron to f5

mitaka-eol
----------

* Updated from global requirements
* Update requirement for horizon in stable/mitaka

1.0.0
-----

* Move to post versioning
* Add release note for LBaaS v2 dashboard work in Mitaka
* Do not allow deleting a listener that has a default pool
* Hide load balancers v2 panel if not enabled
* Add the Angular LBaaS V2 'Update Member List' Table Action
* Reload route after performing actions
* Add trailing slash to all URLs calling the REST API
* Fix gate scripts to use script dir for sourcing commons
* Add action for updating pool member weight
* Add edit health monitor action
* Add create health monitor action
* Add health monitor delete action
* Add the Angular LBaaS V2 'Edit Pool' Workflow
* Add the Angular LBaaS V2 'Create Pool' workflow
* Add pool delete action
* Handle missing key-manager service in workflow
* Add listener delete actions
* Add the Angular LBaaS V2 'Create Listener' workflow
* Update styles to match latest horizon
* Add associate and disassociate floating IP actions
* Add delete load balancer actions
* Make workflow enforce all required fields
* Clean up create load balancer workflow
* Add support for TERMINATED_HTTPS protocol
* Add the angular LBaaS V2 Edit Action for Listeners
* Remove page header from pages on tabs
* Adding integration test support
* Update install instruction for LBaaS v2 dashboard
* Use the horizon limit filter for connection limit
* Add general load balancer service
* Use static file auto-discovery
* Update URL routing
* Updated from global requirements
* Support external members when creating load balancer
* Add the angular LBaaS V2 health monitor detail page
* Add the angular LBaaS V2 members table and detail pages
* Add the angular LBaaS V2 pools detail page
* Fix typo in enable file
* Add angular LBaaS V2 Listeners table and detail pages
* Add support for editing a load balancer
* Make create load balancer workflow generic
* Add monitor tab to create load balancer workflow
* Update npm dependencies
* Add members tab to create load balancer workflow
* Add listener and pool to create load balancer workflow
* Include javascript lint and unit test in tox
* Add initial workflow for creating load balancer
* devstack plugin for neutron-lbaas-dashboard
* Add routing and load balancer detail page
* Add angular Load Balancers V2 panel
* Initial commit for horizon lbaas v2 dashboard
* Make sure django env var is set
* Add requirement for testing
* Open Mitaka development
* Change ignore-errors to ignore_errors
* Initial cookiecutter commit
* Added .gitreview


f5_lbaas_dashboard/tests/settings.py
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

import importlib
import os
import six

from horizon.test.settings import *  # noqa
from horizon.utils import secret_key
from openstack_dashboard import exceptions


DEBUG = True
TEMPLATE_DEBUG = DEBUG

TEST_DIR = os.path.dirname(os.path.abspath(__file__))
ROOT_PATH = os.path.abspath(os.path.join(TEST_DIR, ".."))

MEDIA_ROOT = os.path.abspath(os.path.join(ROOT_PATH, '..', 'media'))
MEDIA_URL = '/media/'
STATIC_ROOT = os.path.abspath(os.path.join(ROOT_PATH, '..', 'static'))
STATIC_URL = '/static/'

SECRET_KEY = secret_key.generate_or_read_from_file(
    os.path.join(TEST_DIR, '.secret_key_store'))
ROOT_URLCONF = 'f5_lbaas_dashboard.tests.urls'
TEMPLATE_DIRS = (
    os.path.join(TEST_DIR, 'templates'),
)

TEMPLATE_CONTEXT_PROCESSORS += (
    'openstack_dashboard.context_processors.openstack',
)

INSTALLED_APPS = (
    'django.contrib.contenttypes',
    'django.contrib.auth',
    'django.contrib.sessions',
    'django.contrib.staticfiles',
    'django.contrib.messages',
    'django.contrib.humanize',
    'django_nose',
    'openstack_auth',
    'compressor',
    'horizon',
    'openstack_dashboard',
    'openstack_dashboard.dashboards',
)

AUTHENTICATION_BACKENDS = ('openstack_auth.backend.KeystoneBackend',)

SITE_BRANDING = 'OpenStack'

HORIZON_CONFIG = {
    "password_validator": {
        "regex": '^.{8,18}$',
        "help_text": "Password must be between 8 and 18 characters."
    },
    'user_home': None,
    'help_url': "http://docs.openstack.org",
    'exceptions': {'recoverable': exceptions.RECOVERABLE,
                   'not_found': exceptions.NOT_FOUND,
                   'unauthorized': exceptions.UNAUTHORIZED},
    'angular_modules': [],
    'js_files': [],
}

# Load the pluggable dashboard settings
from openstack_dashboard.utils import settings
dashboard_module_names = [
    'openstack_dashboard.enabled',
    'openstack_dashboard.local.enabled',
    'f5_lbaas_dashboard.enabled',
]
dashboard_modules = []
# All dashboards must be enabled for the namespace to get registered, which is
# needed by the unit tests.
for module_name in dashboard_module_names:
    module = importlib.import_module(module_name)
    dashboard_modules.append(module)
    for submodule in six.itervalues(settings.import_submodules(module)):
        if getattr(submodule, 'DISABLED', None):
            delattr(submodule, 'DISABLED')
INSTALLED_APPS = list(INSTALLED_APPS)  # Make sure it's mutable
settings.update_dashboards(dashboard_modules, HORIZON_CONFIG, INSTALLED_APPS)


# Set to True to allow users to upload images to glance via Horizon server.
# When enabled, a file form field will appear on the create image form.
# See documentation for deployment considerations.
HORIZON_IMAGES_ALLOW_UPLOAD = True

AVAILABLE_REGIONS = [
    ('http://localhost:5000/v2.0', 'local'),
    ('http://remote:5000/v2.0', 'remote'),
]

OPENSTACK_API_VERSIONS = {
    "identity": 3
}

OPENSTACK_KEYSTONE_URL = "http://localhost:5000/v2.0"
OPENSTACK_KEYSTONE_DEFAULT_ROLE = "_member_"

OPENSTACK_KEYSTONE_MULTIDOMAIN_SUPPORT = True
OPENSTACK_KEYSTONE_DEFAULT_DOMAIN = 'test_domain'

OPENSTACK_KEYSTONE_BACKEND = {
    'name': 'native',
    'can_edit_user': True,
    'can_edit_group': True,
    'can_edit_project': True,
    'can_edit_domain': True,
    'can_edit_role': True
}

OPENSTACK_CINDER_FEATURES = {
    'enable_backup': True,
}

OPENSTACK_NEUTRON_NETWORK = {
    'enable_lb': False,
    'enable_firewall': False,
    'enable_vpn': False
}

OPENSTACK_HYPERVISOR_FEATURES = {
    'can_set_mount_point': True,

    # NOTE: as of Grizzly this is not yet supported in Nova so enabling this
    # setting will not do anything useful
    'can_encrypt_volumes': False
}

LOGGING['loggers']['openstack_dashboard'] = {
    'handlers': ['test'],
    'propagate': False,
}

LOGGING['loggers']['selenium'] = {
    'handlers': ['test'],
    'propagate': False,
}

LOGGING['loggers']['f5_lbaas_dashboard'] = {
    'handlers': ['test'],
    'propagate': False,
}

SECURITY_GROUP_RULES = {
    'all_tcp': {
        'name': 'ALL TCP',
        'ip_protocol': 'tcp',
        'from_port': '1',
        'to_port': '65535',
    },
    'http': {
        'name': 'HTTP',
        'ip_protocol': 'tcp',
        'from_port': '80',
        'to_port': '80',
    },
}

NOSE_ARGS = ['--nocapture',
             '--nologcapture',
             '--cover-package=openstack_dashboard',
             '--cover-inclusive',
             '--all-modules']

POLICY_FILES_PATH = os.path.join(ROOT_PATH, "conf")
POLICY_FILES = {
    'identity': 'keystone_policy.json',
    'compute': 'nova_policy.json'
}

# The openstack_auth.user.Token object isn't JSON-serializable ATM
SESSION_SERIALIZER = 'django.contrib.sessions.serializers.PickleSerializer'


f5_lbaas_dashboard/tests/test_f5_lbaas_dashboard.py
# -*- coding: utf-8 -*-

# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

"""
test_f5_lbaas_dashboard
----------------------------------

Tests for `f5_lbaas_dashboard` module.
"""

from f5_lbaas_dashboard.tests import base


class TestNeutron_lbaas_dashboard(base.TestCase):

    def test_something(self):
        pass


f5_lbaas_dashboard/dashboards/project/loadbalancersv2/views.py
#    Copyright 2015, eBay Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

import logging

from django.core.urlresolvers import reverse
from django.core.urlresolvers import reverse_lazy
from django.utils.translation import ugettext_lazy as _

from horizon import exceptions
from horizon import tables
from horizon import tabs
from horizon import workflows

from f5_lbaas_dashboard import api

from .tables import LoadBalancersTable  # noqa
from .tabs import LoadBalancerDetailTabs  # noqa
from .workflows import LaunchLoadBalancer  # noqa
from .workflows import UpdateLoadBalancer  # noqa


LOG = logging.getLogger(__name__)


class IndexView(tables.DataTableView):
    table_class = LoadBalancersTable
    template_name = 'project/loadbalancersv2/index.html'

    def get_data(self):
        pools = []
        try:
            pools = api.lbaasv2.list_loadbalancers(self.request)
        except Exception:
            exceptions.handle(self.request,
                              _('Unable to retrieve pools list.'))
        return pools


class LaunchLoadBalancerView(workflows.WorkflowView):
    workflow_class = LaunchLoadBalancer
    template_name = "project/loadbalancersv2/launch.html"

    def get_initial(self):
        initial = super(LaunchLoadBalancerView, self).get_initial()
        initial['project_id'] = self.request.user.tenant_id
        initial['user_id'] = self.request.user.id
        return initial


class UpdateView(workflows.WorkflowView):
    workflow_class = UpdateLoadBalancer
    template_name = 'project/loadbalancersv2/update.html'
    success_url = reverse_lazy("horizon:project:loadbalancersv2:index")

    def get_context_data(self, **kwargs):
        context = super(UpdateView, self).get_context_data(**kwargs)
        context["loadbalancer_id"] = self.kwargs['loadbalancer_id']
        return context

    def get_object(self, *args, **kwargs):
        if not hasattr(self, "_object"):
            loadbalancer_id = self.kwargs['loadbalancer_id']
            LOG.info("DEBUGGING - loadbalancer_id=%s" % loadbalancer_id)
            try:
                vip = api.lbaasv2.show_loadbalancer(self.request,
                                                    loadbalancer_id)
                self._object = vip.readable(self.request)
            except Exception as e:
                redirect = reverse("horizon:project:loadbalancersv2:index")
                msg = _('Unable to retrieve load balancer details. %s')\
                    % e.message
                exceptions.handle(self.request, msg, redirect=redirect)
        return self._object

    def get_initial(self):
        initial = super(UpdateView, self).get_initial()

        initial.update({
            'loadbalancer_id': self.kwargs['loadbalancer_id'],
            'address': getattr(self.get_object(), 'address', ''),
            'name': getattr(self.get_object(), 'name', ''),
            'description': getattr(self.get_object(), 'description', ''),
            'lb_method': getattr(self.get_object(), 'lb_method', ''),
            'monitor': getattr(self.get_object(), 'monitor', ''),
            'interval': getattr(self.get_object(), 'interval', ''),
            'timeout': getattr(self.get_object(), 'timeout', ''),
            'send': getattr(self.get_object(), 'send', ''),
            'receive': getattr(self.get_object(), 'receive', ''),

            'source_type': getattr(self.get_object(), 'protocol', ''),
            'http': getattr(self.get_object(), 'port', ''),
            'https': getattr(self.get_object(), 'port', ''),
            'instance_port': getattr(self.get_object(), 'instance_port', ''),
            'selected_members': getattr(self.get_object(), 'members', ''),
            'cert_name': getattr(self.get_object(), 'cert_name', ''),
            'cert': getattr(self.get_object(), 'cert', ''),
            'private_key': getattr(self.get_object(), 'private_key', ''),
            'chain_cert': getattr(self.get_object(), 'chain_cert', ''),
            'enabled': getattr(self.get_object(), 'enabled', ''),
            'use_common_cert': getattr(self.get_object(),
                                       'use_common_cert', ''),
            'port': getattr(self.get_object(), 'port', ''),
        })
        return initial


class DetailView(tabs.TabView):
    tab_group_class = LoadBalancerDetailTabs
    template_name = 'project/loadbalancersv2/detail.html'

    def get_context_data(self, **kwargs):
        context = super(DetailView, self).get_context_data(**kwargs)
        context["loadbalancer"] = self.get_data()
        return context

    def get_data(self):
        if not hasattr(self, "_loadbalancer"):
            try:
                loadbalancer_id = self.kwargs['loadbalancer_id']
                loadbalancer = api.lbaasv2.show_loadbalancer(self.request,
                                                             loadbalancer_id)
            except Exception:
                redirect = reverse('horizon:project:loadbalancersv2:index')
                exceptions.handle(self.request,
                                  _('Unable to retrieve details for '
                                    'loadbalancer "%s".') % loadbalancer_id,
                                  redirect=redirect)
            self._loadbalancer = loadbalancer
        return self._loadbalancer.readable(self.request)

    def get_tabs(self, request, *args, **kwargs):
        loadbalancer = self.get_data()
        return self.tab_group_class(request, loadbalancer=loadbalancer,
                                    **kwargs)


f5_lbaas_dashboard/dashboards/project/ngloadbalancersv2/urls.py
#    (c) Copyright 2015 Hewlett-Packard Development Company, L.P.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from django.conf.urls import patterns
from django.conf.urls import url

from f5_lbaas_dashboard.dashboards.project.ngloadbalancersv2 import views


urlpatterns = patterns(
    'f5_lbaas_dashboard.dashboards.project.ngloadbalancersv2.views',
    url('', views.IndexView.as_view(), name='index'),
)


f5_lbaas_dashboard/dashboards/project/loadbalancersv2/workflows/create_lb.py
#    Copyright 2015, eBay Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

import logging

from django.utils.translation import ugettext_lazy as _
from horizon import exceptions
from horizon import forms
from horizon import workflows

from openstack_dashboard.api import nova

from f5_lbaas_dashboard import api

LOG = logging.getLogger(__name__)
__create_new__ = "Create New"


class SetLBDetailsAction(workflows.Action):

    address = forms.ChoiceField(label=_("IP"),
                                help_text=_("Select from existing VIP IPs"))

    name = forms.CharField(max_length=80, label=_("Name"),
                           required=True)

    description = forms.CharField(widget=forms.Textarea(attrs={'rows': 2}),
                                  label=_(
                                  "Load Balancer Description"),
                                  required=False,
                                  help_text=_("Provide Load Balancer "
                                              "Description."))

    all_vips = None
    is_update = False

    LOAD_BALANCING_CHOICES = (
        ("RoundRobin", _("Round Robin")),
        ("LeastConnection", _("Least Connection")),
        ("LeastSessions", _("Least Sessions"))
    )
    lb_method = forms.ChoiceField(label=_("Load Balancing Method"),
                                  choices=LOAD_BALANCING_CHOICES)

    PROTOCOL_CHOICES = (
        ("HTTP", _("HTTP")),
        ("HTTPS", _("HTTPS")),
        ("TCP", _("TCP")),
        ("SSL", _("SSL")),
    )

    protocol_type = forms.ChoiceField(
        label=_("LB Protocol"), choices=PROTOCOL_CHOICES)

    port = forms.IntegerField(label=_("LB Port"),
                              required=False,
                              min_value=1,
                              max_value=65535,
                              help_text=_("LB Port on which "
                                          "LB is listening."))

    instance_port = forms.IntegerField(label=_("Instance Port"),
                                       required=False,
                                       min_value=1,
                                       max_value=65535,
                                       help_text=_("Instance Port on which "
                                                   "service is running."))

    def __init__(self, request, *args, **kwargs):
        super(SetLBDetailsAction, self).__init__(request, *args, **kwargs)
        self.all_vips = []
        try:
            # todo - this should be obtained in view via an initial method
            self.all_vips = api.lbaasv2.list_loadbalancers(request)
        except Exception:
            pass

        if len(self.fields['address'].choices) == 0:
            del self.fields['address']

    class Meta(object):
        name = _("LB Details")
        help_text_template = ("project/loadbalancersv2/_launch_lb_help.html")

    def clean(self):
        cleaned_data = super(SetLBDetailsAction, self).clean()

        lb_method = cleaned_data['lb_method']
        if not (lb_method == 'RoundRobin'
                or lb_method == 'LeastConnection'
                or lb_method == 'LeastSessions'):
                raise forms.ValidationError(_("Please select an option for "
                                              "the load balancing method."))

        if not self.is_update:
            all_vips = self.all_vips
            ipPortCombo = []
            for vip in all_vips:
                vip = vip.readable()
                ipPortCombo.append('%s:%s' % (vip.address, vip.port))

            data = self.data
            if 'address' in data \
                    and data['address'] != 'new' \
                    and data['address'] != '':
                address = data['address']
                selected_lb_port = data['port']
                selected_ip_port_combo = '%s:%s' % (address.split(':')[0],
                                                    selected_lb_port)
                if selected_ip_port_combo in ipPortCombo:
                    raise forms.ValidationError(_('Requested IP and port '
                                                  'combination already '
                                                  'exists %s ') %
                                                selected_ip_port_combo)

        instance_port = cleaned_data.get('instance_port', None)
        if not instance_port:
            raise forms.ValidationError(
                _('Please provide instance port'))

        return cleaned_data

    def populate_address_choices(self, request, context):
        if self.is_update:
            return []
        try:
            vips = api.lbaasv2.list_loadbalancers(request)
            if len(vips) == 0:
                return []

            distict_ips = set()
            for vip in vips:
                vip = vip.readable()
                distict_ips.add(vip.address)

            existing = []
            for vip in vips:
                vip = vip.readable()
                if vip.address in distict_ips:
                    item = ("%s:%s:%s" %
                            (vip.address, vip.name, 443),
                            "%s" % vip.address)
                    existing.append(item)
                    distict_ips.remove(vip.address)

            vip_list = []
            if len(existing) > 0:
                vip_list.append(('new', __create_new__))
                vip_list.append(('Select Existing', existing))
            return vip_list

        except Exception:
            exceptions.handle(request,
                              _('Unable to retrieve vips.'))
            return []

    def get_help_text(self):
        extra = {}
        return super(SetLBDetailsAction, self).get_help_text(extra)


class SetLBDetails(workflows.Step):
    action_class = SetLBDetailsAction
    contributes = ("name", "description", "lb_method", "protocol_type", "port",
                   "source_id", "instance_port", "address", "monitor")

    def contribute(self, data, context):
        context = super(SetLBDetails, self).contribute(data, context)
        return context

    template_name = "project/loadbalancersv2/launch_lb.html"


class UploadSSLAction(workflows.Action):
    update_cert = forms.BooleanField(label='Update SSL Certificate',
                                     required=False,
                                     widget=forms.HiddenInput())

    cert_name = forms.CharField(max_length=80,
                                label=_("Certificate Name"),
                                required=False)

    cert = forms.CharField(widget=forms.Textarea(attrs={'rows': 3}),
                           label=_("Certificate"),
                           required=False,
                           help_text=_("Certificate"))

    private_key = forms.CharField(widget=forms.Textarea(attrs={'rows': 3}),
                                  label=_("Private Key"),
                                  required=False,
                                  help_text=_("Private Key"))

    chain_cert = forms.CharField(widget=forms.Textarea(attrs={'rows': 3}),
                                 label=_("Certificate Chain (Optional)"),
                                 required=False,
                                 help_text=_("Intermediate Chain"
                                             " Certificates"))

    def clean(self):
        cleaned_data = super(UploadSSLAction, self).clean()
        data = self.data
        protocol = data.get('source_type')
        if protocol == 'HTTPS':
            use_common_cert = data.get('use_common_cert')
            if not use_common_cert:
                # check to see if ssl cert is provided
                cert_name = data.get('cert_name')
                cert = data.get('cert')
                private_key = data.get('private_key')

                if (not cert_name) \
                    or (not cert) \
                        or (not private_key):
                    raise forms.ValidationError(
                        _('Please provide all certificate parameters.'))
        return cleaned_data

    class Meta(object):
        name = _("SSL Certificate")
        help_text_template = ("project/loadbalancersv2/_ssl_cert_help.html")


class UploadSSLStep(workflows.Step):
    action_class = UploadSSLAction
    contributes = ("cert_name", "cert",
                   "private_key", "chain_cert", 'use_common_cert')
    template_name = "project/loadbalancersv2/ssl_cert.html"

    def contribute(self, data, context):
        post = self.workflow.request.POST
        context['cert_name'] = post['cert_name'] if 'cert_name' in post else ''
        context['cert'] = post['cert'] if 'cert' in post else ''
        context['private_key'] = post[
            'private_key'] if 'private_key' in post else ''
        context['chain_cert'] = post[
            'chain_cert'] if 'chain_cert' in post else ''
        context['use_common_cert'] = post[
            'use_common_cert'] if 'use_common_cert' in post else ''
        return context


class SelectInstancesAction(workflows.MembershipAction):
    instance_details = {}

    def __init__(self, request, *args, **kwargs):
        super(SelectInstancesAction, self).__init__(request, *args, **kwargs)
        err_msg = _('Unable to retrieve members list. '
                    'Please try again later.')

        default_role_field_name = self.get_default_role_field_name()
        self.fields[default_role_field_name] = forms.CharField(required=False,
                                                               label='')
        self.fields[default_role_field_name].initial = 'member'

        role_member_field_name = self.get_member_field_name('member')
        self.fields[role_member_field_name] = forms.MultipleChoiceField(
            required=False, label='')

        # Get list of available instances
        all_instances = []
        try:
            all_instances, has_more_data = nova.server_list(request)
        except Exception:
            exceptions.handle(request, err_msg)

        available_instances = []
        for instance in all_instances:
            # skip shutoff instances
            # if instance.status == 'SHUTOFF':
            #     continue
            instance_ip = self.get_ip(instance)
            # skip instances which has no network
            if not instance_ip:
                continue
            key = instance_ip
            value = instance.name + ' (' + self.get_ip(instance) + ')'
            available_instances.append((key, value))
            self.instance_details[instance_ip] = (instance.name, instance.id)

        self.fields[self.get_member_field_name('member')].\
            choices = available_instances

    def get_ip(self, instance):
        ipaddress = None
        for networks in instance.addresses.itervalues():
            for ip in networks:
                # only one IP present
                ipaddress = ip
                break
        if ipaddress is not None:
            addr = ipaddress["addr"]
        else:
            addr = None  # '10.10.10.10'
        return addr

    def clean(self):
        cleaned_data = super(SelectInstancesAction, self).clean()
        members = cleaned_data.get(self.get_member_field_name('member'), None)
        if not members:
            raise forms.ValidationError(
                _('Please select at least one member'))
        return cleaned_data

    class Meta(object):
        name = _("Instances")
        slug = "select_instances"


class SelectInstancesStep(workflows.UpdateMembersStep):
    action_class = SelectInstancesAction
    help_text = _("Please select a list of instances that should handle"
                  " traffic for this target load balancer. All instances "
                  "must reside in the same Project as the target load "
                  "balancer.")
    available_list_title = _("All Instances")
    members_list_title = _("Selected Instances")
    no_available_text = _("No instances found.")
    no_members_text = _("No members enabled.")
    show_roles = False
    contributes = (
        "wanted_members", "instances_details", "monitor", "instance_port")
    template_name = "horizon/common/_workflow_step_update_members.html"

    def contribute(self, data, context):
        request = self.workflow.request
        if data:
            context["wanted_members"] = request.POST.getlist(
                self.get_member_field_name('member'))
            context["instances_details"] = self.action.instance_details
            context["monitor"] = request.POST.get("monitor")
            context["instance_port"] = request.POST.get("instance_port")
        return context


class SelectMonitorAction(workflows.Action):
    MONITOR_CHOICES = (
        ("tcp", _("TCP")),
        ("ping", _("PING")),
        ("http", _("HTTP")),
    )
    monitor = forms.ChoiceField(label=_("Monitor"),
                                choices=MONITOR_CHOICES)

    interval = forms.IntegerField(label=_("Health Check Interval"
                                          " (in seconds)"),
                                  required=False,
                                  min_value=1,
                                  max_value=600,
                                  help_text=_("Health Check Interval"
                                              " (in seconds)"))

    timeout = forms.IntegerField(label=_("Retry count before markdown"),
                                 required=False,
                                 min_value=1,
                                 max_value=100,
                                 help_text=_("Number of times health check "
                                             "should be attempted before "
                                             "marking down a member"))

    send = forms.CharField(widget=forms.Textarea(attrs={'rows': 1}),
                           label=_("Send String"),
                           required=False,
                           help_text=_("Send String"))

    receive = forms.CharField(widget=forms.Textarea(attrs={'rows': 1}),
                              label=_("Receive String"),
                              required=False,
                              help_text=_("Receive String"))

    class Meta(object):
        name = _("Monitor")
        help_text_template = ("project/loadbalancersv2/_monitor_help.html")


class SelectMonitorStep(workflows.Step):
    action_class = SelectMonitorAction
    contributes = ("monitor", "interval", "timeout", "send", "receive")
    template_name = "project/loadbalancersv2/_monitor_create.html"

    def contribute(self, data, context):
        post = self.workflow.request.POST

        context['interval'] = post['interval'] if 'interval' in post else ''
        context['timeout'] = post['timeout'] if 'timeout' in post else ''
        context['send'] = post['send'] if 'send' in post else ''
        context['receive'] = post['receive'] if 'receive' in post else ''
        return context


class LaunchLoadBalancer(workflows.Workflow):
    slug = "launch_loadbalancer"
    name = _("Launch Load Balancer")
    finalize_button_name = _("Launch")
    success_message = _('Launched %(count)s named "%(name)s".')
    failure_message = _('Unable to launch %(count)s named "%(name)s".')
    success_url = "horizon:project:loadbalancersv2:index"
    default_steps = (SetLBDetails,
                     UploadSSLStep,
                     SelectMonitorStep,
                     SelectInstancesStep,
                     )
    attrs = {'data-help-text': 'LB creation may take a few minutes'}

    def format_status_message(self, message):
        name = self.context.get('name', 'unknown loadbalancer')
        count = self.context.get('count', 1)
        if int(count) > 1:
            return message % {"count": _("%s loadbalancers") % count,
                              "name": name}
        else:
            return message % {"count": _("loadbalancer"), "name": name}

    def handle(self, request, context):
        try:
            protocol = context['source_type']
            address = context['address']
            if not address\
                    or address == "new":
                address = ''
            else:
                tokens = address.split(':')
                address = tokens[0]

            api.lbaasv2.\
                create_loadbalancer_full(request,
                                         address=address,
                                         name=context['name'],
                                         description=context['description'],
                                         lb_method=context['lb_method'],
                                         monitor=context['monitor'],
                                         protocol=protocol,
                                         port=context[protocol],
                                         instance_port=context['instance_port'],  # noqa
                                         wanted_members=context['wanted_members'],  # noqa
                                         instances_details=context['instances_details'],  # noqa
                                         cert_name=context['cert_name'],
                                         cert=context['cert'],
                                         private_key=context['private_key'],
                                         chain_cert=context['chain_cert'],
                                         use_common_cert=True if
                                            context['use_common_cert'] == 'on'
                                            else False,
                                         interval=context['interval'],
                                         timeout=context['timeout'],
                                         send=context['send'],
                                         receive=context['receive'],
                                         )
            return True
        except Exception as e:
            exceptions.handle(request, e.message, ignore=False)
            return False
